Added Swim Speed Attribute to EntityLivingBase

LetterShapedGlyphs opened the pull request:
Gives the EntityPlayer the attribute and applies it with other
vanilla/forge attributes.
Provides local implementations of EntityLivingBase's handleJumpWater()
and Entity's moveRelative to prevent code from running for every entity.

LetterShapedGlyphs committed:
Added Swim Speed Attribute to EntityPlayer

Gives the EntityPlayer the attribute and applies it with other
vanilla/forge attributes.
Provides local implementations of EntityLivingBase's handleJumpWater()
and Entity's moveRelative to prevent code from running for every entity.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/MinecraftForge/MinecraftForge?pullRequest=4990) <br/>All committers have signed the CLA.

LetterShapedGlyphs commented:
Previous implementations of buffing swimming speed involve onTick effects directly modifying motionX Y and Z, which accelerates external forces while in water in addition to being extremely inaccurate (multiplying motionX by 1.2 results in moving approximately 2-3 times faster as the effect compounds until the friction matches the multiplier)

By implementing it as an attribute, mods can choose between additive or multiplicative scaling of the attribute, ensuring they work together without becoming uncontrollable. Additionally, this simplifies the overhead of mods which try to modify movement in water, allowing armor or baubles to apply the attribute while worn instead of responding to onTick once per mod. Lastly, this change has no affect on vanilla mechanics as the default attribute is 1, and thus duplicates existing water movement handling.

This also serves as a jumping off point for potential future modifications around swimming which might become more relevant in the 1.13 aquatic update.

williewillus commented on src/test/java/net/minecraftforge/debug/entity/player/SwimSpeedAttributeTest.java:
```diff
@@ -0,0 +1,80 @@
+        }
+    }
+
+    public static class ExtendedReachPlate extends ItemArmor
```
Rename?

LetterShapedGlyphs replied:
Woooops. Looks like my shameless rip of the test mod has been revealed!

williewillus commented on patches/minecraft/net/minecraft/entity/player/EntityPlayer.java.patch:
```diff
@@ -9,35 +9,37 @@
 +    protected java.util.HashMap<Integer, Boolean> spawnForcedMap = new java.util.HashMap<Integer, Boolean>();
 +    public float eyeHeight = this.getDefaultEyeHeight();
 +    public static final net.minecraft.entity.ai.attributes.IAttribute REACH_DISTANCE = new net.minecraft.entity.ai.attributes.RangedAttribute(null, "generic.reachDistance", 5.0D, 0.0D, 1024.0D).func_111112_a(true);
++    public static final net.minecraft.entity.ai.attributes.IAttribute SWIM_SPEED = new net.minecraft.entity.ai.attributes.RangedAttribute(null, "forge.swimSpeed", 1.0D, 0.0D, 1024.0D).func_111112_a(true);
```
Why not have this in entitylivingbase? Non players can be affected by this too, I'm pretty sure

LetterShapedGlyphs replied:
I honestly hadn't considered that until after I submitted the pull request. I think you're right that there's no reason NOT to have it there as the code itself doesn't add any additional overhead for it's execution.

LetterShapedGlyphs committed:
Migrated Attribute to EntityLivingBase

No longer requires duplicating handleJumpWater, only modifying, and
allows for non-player entities to implement new Attribute.

LetterShapedGlyphs commented:
As a note: Currently I only have the Swim_Speed attribute affecting water based swimming. The only mod I know of that modifies lava-based swimming is Astral Sorcery, but I'm not sure how warranted/useful it is to either A) have Swim_Speed apply to lava as well, or B) add an attribute specifically for swimming in Lava.

Both options require only minimal changes from the most recent commit, but I feel like "Add an attribute for everything" is not necessarily the correct mindset. That said, I'm not sure that it would have any sort of negative impact to have a second and it would allow for slightly more flexibility.

MajorTuvok commented:
I'd say you'd have to add an "attribute for everything", so that other mods could start using it with their custom fluids?
If you don't intend to support custom fluids, I'd agree that an extra lava attribute would be good...
Though you could probably add an Comment, that mods should simply use the water attribute...

LetterShapedGlyphs commented:
Amusingly enough, forge's custom fluids already cause Entity.inWater() to return true (If they're set up the way the comments suggest). It is actually only Lava that has it's own special material check with world.isMaterialInBB(). In fact, given that modded fluids can easily override "onEntityCollidedWithBlock", it would be very simple to apply temporary attribute modifiers through potion effects while immersed in them. I'll attach some comments outlining that process and update it later today.

Any other things I should note/change while I'm at it?

Alpvax commented:
I don't think lava should be special cased, I would rather the lava just applied an attribute modifier.

mezz commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,63 @@
++        this.field_70181_x += 0.03999999910593033D * this.func_110148_a(SWIM_SPEED).func_111126_e();
+     }
+-
++    
```
please avoid unnecessary whitespace changes in patches

mezz commented on patches/minecraft/net/minecraft/entity/EntityLivingBase.java.patch:
```diff
@@ -247,15 +263,63 @@
+         return 0.8F;
+     }
++    
++    public void func_191958_b(float strafe, float up, float forward, float friction)
```
please add an `@Override` annotation and move this to the end of the file after a comment like this:
`/* ======================================== FORGE START ======================================== */`

LetterShapedGlyphs replied:
When I moved from EntityPlayer to EntityLivingBase I didn't realize there was a Forge section already. I have moved the overridden method there and added the annotation.

mezz commented on src/test/java/net/minecraftforge/debug/entity/player/SwimSpeedAttributeTest.java:
```diff
@@ -0,0 +1,80 @@
+/*
+ * Minecraft Forge
+ * Copyright (c) 2016.
```
Please update this to follow the new copyright format that has recently rolled out.

mezz commented on src/test/java/net/minecraftforge/debug/entity/player/SwimSpeedAttributeTest.java:
```diff
@@ -0,0 +1,80 @@
+            return attribs;
+        }
+    }
+}
```
files should end with a newline

LetterShapedGlyphs replied:
I think I fixed this one but the .diff that git was showing me makes me unsure.

Edit: Nevermind, I was just misreading it. Fixed.

mezz reviewed (changes requested):
Mostly looks good, here's some comments for review

LetterShapedGlyphs committed:
Applied to Lava & Requested Changes

Applied swimspeed to lava: SwimSpeed now affects all fluids. (Forge
Fluids count as water)
Fixed formatting errors.

LetterShapedGlyphs committed:
Patch cleanup

Moved handleJumpWater and handleJumpLava back to where they are supposed
to be to minimize .patch changes.

LetterShapedGlyphs commented:
Made the changes and made the attribute apply to lava as well. (Forge fluids inherit from water, so I thought it'd be weird for lava to be the only fluid not affected at all.)

mezz commented:
Please run `gradle updateLicenses` to fix the license headers on your PR.

LetterShapedGlyphs committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.12.x' into SwimSpeedAttribute

LetterShapedGlyphs commented:
The gradle script didn't have the updateLicenses task so I rebased, after which running the script didn't generate any diffs, so let me know if that didn't fix it.

mezz commented:
All the build checks passed so it looks like you're fine 👍

LetterShapedGlyphs committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.12.x' into SwimSpeedAttribute

LetterShapedGlyphs commented:
So it looks like someone else patched EntityLivingBase while this was pending and... I have no clue how to resolve that conflict because it's within the patch file itself and I don't fully understand the syntax involved in the conflicting lines, or how they're even conflicting in the first place.

LetterShapedGlyphs committed:
Patch conflict resolution?

LetterShapedGlyphs committed:
Conflict Resolution attempt 2

Re-add base changes to a fresh clone

karthas077 committed:
Merge branch '1.12.x' into SwimSpeedAttribute

LetterShapedGlyphs commented:
Re-added my changes to a fresh clone and regenerated the patch so I could be confident about which line in the conflict was correct.

mezz commented:
Fixing merge conflicts in patches is very confusing, you took the right approach.
